### DISCUSSION TITLE
Migration from Eclipse to VSCode and Epsilon compatibility

### DISCUSSION THREAD

**Discussion by ichxorya** [eclipse-langium/langium#1212] [Q&A]
As the title mentioned, my team has a plan to migrate our projects from Eclipse to VSCode, with the new correspondence technology (Langium for Xtext, ect.)

I would like to ask if anyone in the community has experience with this problem, as Langium is new and we don't have much resources to learn from.

There is also Eclipse Epsilon (https://eclipse.dev/epsilon/) that may fit the purpose of MDD components (replacement for Acceleo, ATL...), yet we also new in this field, too. Have anyone used this with Langium, do you have any advices?

Thank you for reading this!

**Comment by cdietrich** [comment]
as i have done a ton of Xtext in the past and now use langium i can say it heavily depends on which subset of Xtext you are using if it works smooth or you will face problems / lacking features

**Comment by msujew** [comment]
At [TypeFox](https://www.typefox.io/) we have performed a few migrations from Xtext to Langium. Obviously development on Langium is driven quite a bit from such work. So far, we haven't found larger issues with that.

I know of a few others that have migrated from Xtext to Langium. Maybe @ballcoach12 can chime in on that.

The main differences between Langium and Xtext from a feature perspective are mostly related to model serialization, reflection (see https://github.com/eclipse-langium/langium/issues/1209) and the general ecosystem (i.e. there are no additional libraries for Langium yet that add features on top, such as xsemantics for Xtext).

**Comment by ballcoach12** [comment]
We have migrated three languages to Langium, and so far, I have not found any features that we were using in Xtext that we cannot implement in Langium.  

Langium's grammar syntax is almost identical, to Xtext and has the added benefit of not forcing you to deal with ECore.  There is no corollary to XCore in Langium, but that isn't a limitation, in my opinion.  Also, Typescript's support for union types makes things more streamlined when generating AST from grammar, in my opinion, which is a factor that leans heavily in Langium's favor.

The only thing that Langium doesn't have that we were using in Xtext were grammar enums, but that is pretty easy to work around using constants in utility classes, so we don't really miss them all that much.  

My experience with VSCode extensions is that they are far easier to develop and manage than Eclipse plugins.  We are delivering all our editors in a browser context these days, and Langium makes that super easy.  The next release should make it even easier with the updated examples for using `monaco-editor-wrapper` that the TypeFox folks are working on.

In summary, I highly recommend Langium.  To me, it's less confining than Xtext, and there is no legacy software cruft that comes with it.

Hope this helps.
